### ISSUE TITLE
PDE freezes after switching into the Android mode

### ISSUE THREAD

**Issue by codeanticode** [processing/processing-android#209]
Observed on PR3 under OS X 10.10

**Comment by codeanticode** [comment]
After waiting for a while, the PDE is responsive again, but the following error shows up in the console (3.0.2):

```
Exception in thread "AWT-EventQueue-0" java.lang.OutOfMemoryError: GC overhead limit exceeded
	at java.util.regex.Matcher.<init>(Matcher.java:225)
	at java.util.regex.Pattern.matcher(Pattern.java:1093)
	at com.google.classpath.RegExpResourceFilter.match(RegExpResourceFilter.java:37)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:50)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:32)
	at com.google.classpath.DirectoryClassPath.findResources(DirectoryClassPath.java:90)
	at com.google.classpath.CompositeClassPath.findResources(CompositeClassPath.java:78)
	at processing.mode.java.pdex.ASTGenerator.getSuggestImports(ASTGenerator.java:3288)
```

Same thing happens with latest revision from Processing (0249), the error is slightly different:

```
Exception in thread "AWT-EventQueue-0" java.lang.OutOfMemoryError: GC overhead limit exceeded
	at java.util.regex.Matcher.<init>(Matcher.java:225)
	at java.util.regex.Pattern.matcher(Pattern.java:1093)
	at com.google.classpath.RegExpResourceFilter.match(RegExpResourceFilter.java:37)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:50)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:46)
	at com.google.classpath.ResourceFinder.findResources(ResourceFinder.java:32)
	at com.google.classpath.DirectoryClassPath.findResources(DirectoryClassPath.java:90)
	at com.google.classpath.CompositeClassPath.findResources(CompositeClassPath.java:78)
	at processing.mode.java.pdex.ASTGenerator.getSuggestImports(ASTGenerator.java:3288)
```

**Comment by Manindra29** [comment]
This error is from the [google classpath explorer library](https://github.com/frankslin/classpath-explorer) which is used for providing import suggestions. It's running out of memory while performing a regex match to find all possible classes that match some class in the classpath. The Android SDK definitely adds a lot more classes to its search path on top of the default pde classpath classes. I'll investigate further and report back.

**Comment by codeanticode** [comment]
The only jar in the Android SDK that should be searched for classes to import is platforms-xy/android.jar, and if several platforms are available, only the currently selected one . Is there any way to specific the particular jar to include in the classpath?

**Comment by codeanticode** [comment]
Just a quick update: forcing ASTGenerator.getSuggestImports() to return null regardless of the situation does in fact eliminate the UI hanging and the OutOfMemoryError. So we clearly need a way to limit the extent of search by the classpath explorer...

**Comment by codeanticode** [comment]
A temporary workaround is to disable the "Suggest import statements" option in the Preferences.

**Comment by codeanticode** [comment]
@JakubValtar adding you to this discussion. As Manindra pointed out above, this issue is due to the classpath explorer library, used by the suggest imports features. The Android SDK has too many classes, and this leads to the out-of-memory error.

**Comment by JakubValtar** [comment]
@codeanticode Update processing to latest and try again. I did error checker overhaul (merged yesterday) and it works great for me with -Xmx256m. Problem might be that previously whole processing app class path was added to search path, which is certainly not what we wanted and for example IDEs are adding tons of stuff to class path, so some people may have experienced problems with having too much stuff there.

**Comment by JakubValtar** [comment]
@codeanticode I fixed memory leak in AndroidEditor which was preventing closed editors to be GC'd. Might solve this. I can run more than 10 editors within 256 MB. See https://github.com/processing/processing-android/pull/215

**Comment by codeanticode** [comment]
Awesome, thanks a lot! I I just run the mode using the PDE after you checker overhaul, and didn't observe the freeze so far. I will now look into the PR.

**Comment by codeanticode** [comment]
The freeze is solved in 3.1, when using 3.0.x, disable the "Suggest import statements" option in the Preferences as a workaround.

**Comment by JakubValtar** [comment]
Glad it helped!
